device_twins_json_object/main.c

```diff
@@ -0,0 +1,208 @@
+            key_bool_value = (bool)json_object_get_boolean(root_object, "keyBool");
+            Log_Debug("keyBool = %s\n", key_bool_value? "true": "false");
+        } else {
+            Log_Debug("keyBool not found!\n");
+        }
+
+        // The SampleJsonObject properties should have a "keyInt" key
+        if (json_object_has_value(root_object, "keyInt") != 0) {
+            key_int_value = (int)json_object_get_number(root_object, "keyInt");
+            Log_Debug("keyInt = %d\n", key_int_value);
+        } else {
+            Log_Debug("keyInt not found!\n");
+        }
+
+        // The SampleJsonObject properties should have a "keyFloat" key
+        if (json_object_has_value(root_object, "keyFloat") != 0) {
+            key_float_value = (float)json_object_get_number(root_object, "keyFloat");
+            Log_Debug("keyFloat = %.2f\n", key_float_value);
+        } else {
+            Log_Debug("keyFloat not found!\n");
+        }
+
+        // The SampleJsonObject properties should have a "keyDouble" key
+        if (json_object_has_value(root_object, "keyDouble") != 0) {
+            key_double_value = (double)json_object_get_number(root_object, "keyDouble");
+            Log_Debug("keyDouble = %.4lf\n", key_double_value);
+        } else {
+            Log_Debug("keyDouble not found!\n");
+        }
+
+        // The SampleJsonObject properties should have a "keyString" key
+        if (json_object_has_value(root_object, "keyString") != 0) {
+            strncpy(key_string_value, json_object_get_string(root_object,"keyString"), MAX_STRING_LEN);
+            Log_Debug("keyString = %s\n", key_string_value);
+        } else {
+            Log_Debug("keyString not found!\n");
+        }      
+
+        // When we have nested objects within an object, we get a new JSON_Object pointer to the 
+        // nested object and pull data as above.
+
+        // The SampleJsonObject properties should have a "keyJsonObj" key
+        if (json_object_has_value(root_object, "keyJsonObj") != 0) {
+
+            // Check to see if the object contains a "keyJsonObj" object
+            key_nested_json_obj = json_object_get_object(root_object, "keyJsonObj");
+            if (key_nested_json_obj == NULL) {
+                Log_Debug("keyJsonObj not found\n");
+            }
+
+            // The keyJsonObj properties should have a "nestedKeyInt" key
+            if (json_object_has_value(key_nested_json_obj, "nestedKeyInt") != 0) {
+                key_nested_key_int_value = (int)json_object_get_number(key_nested_json_obj, "nestedKeyInt");
+                Log_Debug("nestedKeyInt = %d\n", key_nested_key_int_value);
+            } else {
+                Log_Debug("nestedKeyInt not found!\n");
+            }
+        } else {
+            Log_Debug("keyJsonObj not found!\n");
+        }      
+
+        // We need to manually build the response JSON
+        
+        // Define the JSON structure
```

Comment:
I'm doing something like this to serialize json objects, thought I'd share
unfortunately, it allocates memory so you have to be careful to free it

JSON_Value *rootValue = json_value_init_object();
JSON_Object *rootObject = json_value_get_object(rootValue);

json_object_set_boolean(rootObject, "keyBool", key_bool_value);
json_object_set_number(rootObject, "keyInt", key_int_value);
json_object_set_number(rootObject, "keyFloat", key_float_value);
json_object_set_number(rootObject, "keyDouble", key_double_value);
json_object_set_string(rootObject, "keyString", key_string_value);

JSON_Value *keyJsonObjValue = json_value_init_object();
JSON_Object *keyJsonObjObject = json_value_get_object(keyJsonObjValue);

json_object_set_number(keyJsonObjObject, "keyInt", key_nested_key_int_value);

json_object_set_object(rootObject, "keyJsonObj", keyJsonObjObject);

char *pSerialized = json_serialize_to_string(rootValue);

strcpy(msgBuffer, pSerialized);

json_free_serialized_string(pSerialized);
json_value_free(rootValue);

Reply:
That's awesome!  If you did not have the JsonObject inside the JsonObject you could also use the helper function: ```bool dx_jsonSerialize(char* buffer, size_t buffer_size, int key_value_pair_count, ...);```  This function takes (Type, Key, Value) triples and constructs JSON for you.

Reply:
I prefer to "roll my own" json because the parson library does not play nice with floats or doubles, you can't specify how many decimal places to place in the value part.

Reply:
i actually modified dx_jsonSerialize to work on DX_JSON_OBJECT
i dont really like the addition (again with allocated memory issues), so will probably revert